pkg/querier/multi_tenant_querier.go

```diff
@@ -136,6 +153,94 @@ func (q *MultiTenantQuerier) Series(ctx context.Context, req *logproto.SeriesReq
 	return logproto.MergeSeriesResponses(responses)
 }
 
+// removeTenantSelector filters the given tenant IDs based on any tenant ID filter the in passed selector.
+func removeTenantSelector(params logql.SelectSampleParams, tenantIDs []string) (map[string]struct{}, syntax.SampleExpr, error) {
+	expr, err := params.Expr()
+	if err != nil {
+		return nil, nil, err
+	}
+	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, tenantIDs, expr.Selector().Matchers()...)
+	expr = replaceMatchers(expr, filteredMatchers)
+	return matchedTenants, expr, nil
+}
+
+// replaceMatchers traverses the passed epxression and replaces all matchers.
+func replaceMatchers(expr syntax.SampleExpr, matchers []*labels.Matcher) syntax.SampleExpr {
```

Comment:
@ssncferreira and @chaudum, I've added you because I saw a mapper in your recent changes. I'm wondering if we should add a visitor pattern to the AST. What do you think?